File: frontend/src/services/types/entities.ts
import type {
  BetGroupId, DrawEntryId, DrawId, DrawSectionId, IsoDate, IsoDateTime,
  ParticipationId, PlayerId, QuestionId, ScoreEntryId, TournamentId, UserId,
} from './ids';
import type {
  AnswerType, BetGroupKind, BetGroupStatus, MatchFormat, ParticipationStatus,
  QuestionFamily, RoundReached, Surface, Tour, TournamentCategory,
  TournamentStatus, TournamentVisibility, TypedQuestionKind,
} from './enums';
import type { QuestionOutcomePayload } from './predictions';

/** Spec 6.1. Password hashes never cross the API boundary. */
export interface User {
  readonly id: UserId;
  readonly email: string;
  readonly displayName: string;
  readonly fullName: string | null;
  readonly isAdmin: boolean;
  readonly isActive: boolean;
  readonly emailVerifiedAt: IsoDateTime | null;
  readonly createdAt: IsoDateTime;
}

/** The global player registry (spec 6.1). */
export interface Player {
  readonly id: PlayerId;
  readonly fullName: string;
  /** ISO 3166-1 alpha-3, e.g. `ESP`. */
  readonly countryCode: string;
  readonly tour: Tour;
}

/**
 * Point values for one game, editable per tournament (decision D10). Structure
 * is fixed; values are not. Editing a profile does not retroactively rescore.
 */
export interface ScoringProfile {
  readonly quarterFinalistPoints: number;
  readonly semiFinalistPoints: number;
  readonly finalistPoints: number;
  readonly championPoints: number;
  /** Points for an underperformer exiting in round 1, 2 and 3 respectively. */
  readonly underperformerPoints: readonly [number, number, number];
  /** Points for a breakout reaching R16, QF, SF, F and winning respectively. */
  readonly breakoutPoints: readonly [number, number, number, number, number];
  readonly featuredMatchWinnerPoints: number;
  readonly featuredMatchSetScorePoints: number;
}

export interface Tournament {
  readonly id: TournamentId;
  readonly name: string;
  readonly category: TournamentCategory;
  readonly surface: Surface;
  readonly location: string;
  readonly startDate: IsoDate;
  readonly endDate: IsoDate;
  readonly signupDeadline: IsoDateTime;
  readonly visibility: TournamentVisibility;
  /** Present only for private games, and only to an admin. */
  readonly joinCode: string | null;
  readonly status: TournamentStatus;
  readonly rulesMarkdown: string;
  readonly scoringProfile: ScoringProfile;
  readonly createdAt: IsoDateTime;
}

/** First-class child of a tournament, one or two per game (decision D6). */
export interface Draw {
  readonly id: DrawId;
  readonly tournamentId: TournamentId;
  readonly tour: Tour;
  readonly drawSize: number;
  /** Sets needed to win a match: 3 for best of five, 2 for best of three. */
  readonly bestOf: MatchFormat;
  readonly officialDrawUrl: string | null;
}

/** One eighth of a draw. Each section feeds exactly one quarter-final slot. */
export interface DrawSection {
  readonly id: DrawSectionId;
  readonly drawId: DrawId;
  /** 1..8. */
  readonly index: number;
}

export interface DrawEntry {
  readonly id: DrawEntryId;
  readonly drawId: DrawId;
  readonly sectionId: DrawSectionId;
  readonly playerId: PlayerId;
  /** Null for an unseeded entrant. */
  readonly seed: number | null;
}

/** A section with its entrants resolved, which is how the UI reads a draw. */
export interface DrawSectionWithEntries {
  readonly section: DrawSection;
  readonly entries: readonly DrawEntryWithPlayer[];
}

export interface DrawEntryWithPlayer {
  readonly entry: DrawEntry;
  readonly player: Player;
  /** From the outcome grid. Null until the admin records it. */
  readonly roundReached: RoundReached | null;
}

export interface Participation {
  readonly id: ParticipationId;
  readonly userId: UserId;
  readonly tournamentId: TournamentId;
  readonly joinedAt: IsoDateTime;
  readonly status: ParticipationStatus;
}

export interface BetGroup {
  readonly id: BetGroupId;
  readonly tournamentId: TournamentId;
  readonly kind: BetGroupKind;
  /** Null for the tournament bet group. */
  readonly round: number | null;
  readonly title: string;
  readonly introMarkdown: string;
  readonly deadline: IsoDateTime;
  readonly status: BetGroupStatus;
}

export interface QuestionChoiceOption {
  readonly id: string;
  readonly label: string;
}

/**
 * Single-table inheritance (spec 6.2): `family` splits typed from generic,
 * `kind` narrows the typed case and is null for a generic question.
 */
export interface Question {
  readonly id: QuestionId;
  readonly betGroupId: BetGroupId;
  /** Null for a question that is not bound to one draw. */
  readonly drawId: DrawId | null;
  readonly family: QuestionFamily;
  readonly kind: TypedQuestionKind | null;
  readonly prompt: string;
  readonly answerType: AnswerType;
  /** Present when `answerType` is CHOICE. */
  readonly options: readonly QuestionChoiceOption[] | null;
  /**
   * Present when `answerType` is MATCH_RESULT: the two players in the tie, in
   * the order they should be shown. A prediction's winner must be one of them.
   */
  readonly matchup: readonly [PlayerId, PlayerId] | null;
  /** Points available, for display. The backend remains the authority. */
  readonly pointsHint: string;
  /** Unexposed per-question override (decision D7). Null in the MVP UI. */
  readonly deadlineOverride: IsoDateTime | null;
  readonly position: number;
}

/** One row of the outcome grid (decision D9). */
export interface OutcomeEntry {
  readonly drawId: DrawId;
  readonly playerId: PlayerId;
  readonly roundReached: RoundReached;
  readonly note: string | null;
}

export interface QuestionOutcome {
  readonly questionId: QuestionId;
  readonly correctAnswer: QuestionOutcomePayload;
  readonly settledAt: IsoDateTime;
  readonly note: string | null;
}

/** Derived data, rebuildable at any time (decision D9, AGENTS.md rule 7). */
export interface ScoreEntry {
  readonly id: ScoreEntryId;
  readonly participationId: ParticipationId;
  readonly questionId: QuestionId;
  readonly points: number;
  /** Human-readable audit line, e.g. "6 of 8 correct". */
  readonly reason: string;
  readonly calculatedAt: IsoDateTime;
}

File: backend/app/domain/entities.py
"""Domain entities (spec 6.1). Plain dataclasses, no framework."""

from __future__ import annotations

from dataclasses import dataclass, field
from datetime import date, datetime

from app.domain.enums import (
    AnswerType,
    BetGroupKind,
    BetGroupStatus,
    ParticipationStatus,
    QuestionFamily,
    RoundReached,
    Surface,
    Tour,
    TournamentCategory,
    TournamentStatus,
    TournamentVisibility,
    TypedQuestionKind,
)
from app.domain.predictions import QuestionOutcomePayload


@dataclass(slots=True)
class User:
    id: str
    email: str
    password_hash: str
    display_name: str
    full_name: str | None
    is_admin: bool
    is_active: bool
    email_verified_at: datetime | None
    created_at: datetime


@dataclass(frozen=True, slots=True)
class Player:
    id: str
    full_name: str
    country_code: str
    tour: Tour


@dataclass(frozen=True, slots=True)
class ScoringProfile:
    """Point values per tournament, editable by the organiser (decision D10)."""

    quarter_finalist_points: int = 1
    semi_finalist_points: int = 2
    finalist_points: int = 3
    champion_points: int = 5
    #: Exit in round 1, 2 and 3 respectively.
    underperformer_points: tuple[int, int, int] = (3, 2, 1)
    #: Reaching R16, QF, SF, F and winning respectively.
    breakout_points: tuple[int, int, int, int, int] = (2, 3, 4, 5, 7)
    featured_match_winner_points: int = 1
    featured_match_set_score_points: int = 1


@dataclass(slots=True)
class Tournament:
    id: str
    name: str
    category: TournamentCategory
    surface: Surface
    location: str
    start_date: date
    end_date: date
    signup_deadline: datetime
    visibility: TournamentVisibility
    join_code: str | None
    status: TournamentStatus
    rules_markdown: str
    created_at: datetime
    scoring_profile: ScoringProfile = field(default_factory=ScoringProfile)


@dataclass(frozen=True, slots=True)
class Draw:
    id: str
    tournament_id: str
    tour: Tour
    draw_size: int
    best_of: int
    official_draw_url: str | None


@dataclass(frozen=True, slots=True)
class DrawSection:
    id: str
    draw_id: str
    index: int


@dataclass(frozen=True, slots=True)
class DrawEntry:
    id: str
    draw_id: str
    section_id: str
    player_id: str
    seed: int | None


@dataclass(frozen=True, slots=True)
class Participation:
    id: str
    user_id: str
    tournament_id: str
    joined_at: datetime
    status: ParticipationStatus


@dataclass(slots=True)
class BetGroup:
    id: str
    tournament_id: str
    kind: BetGroupKind
    round: int | None
    title: str
    intro_markdown: str
    deadline: datetime
    status: BetGroupStatus


@dataclass(frozen=True, slots=True)
class QuestionChoiceOption:
    id: str
    label: str


@dataclass(slots=True)
class Question:
    id: str
    bet_group_id: str
    draw_id: str | None
    family: QuestionFamily
    kind: TypedQuestionKind | None
    prompt: str
    answer_type: AnswerType
    options: tuple[QuestionChoiceOption, ...] | None
    matchup: tuple[str, str] | None
    points_hint: str
    deadline_override: datetime | None
    position: int


@dataclass(slots=True)
class Prediction:
    id: str
    participation_id: str
    question_id: str
    payload: object
    submitted_at: datetime | None
    updated_at: datetime


@dataclass(frozen=True, slots=True)
class OutcomeEntry:
    """One row of the outcome grid (decision D9)."""

    draw_id: str
    player_id: str
    round_reached: RoundReached
    note: str | None


@dataclass(frozen=True, slots=True)
class QuestionOutcome:
    question_id: str
    correct_answer: QuestionOutcomePayload
    settled_at: datetime
    note: str | None


@dataclass(frozen=True, slots=True)
class ScoreEntry:
    """Derived data, rebuildable at any time (AGENTS.md hard rule 7)."""

    id: str
    participation_id: str
    question_id: str
    points: int
    reason: str
    calculated_at: datetime
